### PR TITLE
Port iOS FFI interface to Android

### LINKED ISSUE #81: Port ios_interface to Android
This makes core's functionality usable by Android.

### LINKED ISSUE #84: Replace / update confy
Not working on Android: https://github.com/rust-cli/confy/issues/31

Unless they add support soon, we need something else to manage preferences.

### PR DESCRIPTION
**Pull request by ivnsch**

### CODE REVIEW

File: src/composition_root.rs [hunk 1]
```diff
@@ -69,23 +70,27 @@ fn create_comp_root() -> CompositionRoot<
...
+
+    // TODO confy seems not to work on Android
+    // let preferences = Arc::new(PreferencesImpl {
+    //     // unwrap: "Errors that are returned from this function are I/O related,
+    //     // for example if the writing of the new configuration fails or confy encounters
+    //     // an operating system or environment that it does not support."
+    //     // The config is critical in this app, so it's ok to crash if not available.
+    //     config: RwLock::new(confy::load("coepi")),
+    // });
+    let preferences = Arc::new(PreferencesNoopMock {});
```
Comment (duskoo) [inline src/composition_root.rs]: Suggest we use conditional compilation: PreferencesNoopMock for target_os == android.
That way we dont break the functionality for iOS.

**Review by duskoo — APPROVED** [review]
PreferencesNoopMock might affect iOS

File: src/composition_root.rs (hunk 1 above)
Comment (ivnsch) [reply src/composition_root.rs]: Draft PR! It's still in progress.

File: src/preferences.rs [hunk 2]
```diff
@@ -62,57 +153,51 @@ pub trait Preferences {
...
-        let mut config = self.config.write();
-        config.last_completed_reports_interval = Some(value);
-
-        let res = confy::store("myprefs", *config);
-
-        if let Err(error) = res {
-            error!("Error storing preferences: {:?}", error)
-        }
+    fn set_last_completed_reports_interval(&self, value: ReportsInterval) {
+        let res = serde_json::to_string(&value);
```
Comment (duskoo) [inline src/preferences.rs:169]: @i-schuetz Why conversion to string?

File: src/android/android_interface.rs [hunk 3]
```diff
@@ -153,3 +416,100 @@ fn register_callback_internal(callback: Box<dyn LogCallbackWrapper>) {
         }
     });
 }
+
+// To prefill the JNI array (TODO can this be skipped?)
```
Comment (duskoo) [inline src/android/android_interface.rs:420]: I found the same approach on SO: https://stackoverflow.com/a/16473941

File: src/android/android_interface.rs [hunk 4]
```diff
@@ -27,12 +41,258 @@ pub unsafe extern "C" fn Java_org_coepi_android_api_NativeApi_bootstrapCore(
...
+    )
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Java_org_coepi_android_api_NativeApi_recordTcn(
+    env: JNIEnv,
+    _: JClass,
+    tcn: JString,
+) -> jobject {
+    let tcn_java_str = env.get_string(tcn).unwrap();
```
Comment (duskoo) [inline src/android/android_interface.rs:96]: Should we avoid unwrap?

**Review by duskoo — APPROVED** [review]

File: src/preferences.rs (hunk 2 above)
Comment (ivnsch) [reply src/preferences.rs:169]: Because the sqlite preferences table is (text, text)

File: src/android/android_interface.rs (hunk 4 above)
Comment (ivnsch) [reply src/android/android_interface.rs:96]: Like last time, The error handling improvements will be in a separate PR (see [issue](https://github.com/Co-Epi/app-backend-rust/issues/79). It should only affect this file. I want to finish the app integration ASAP.

File: src/preferences.rs (hunk 2 above)
Comment (ivnsch) [reply src/preferences.rs:169]: Another question would be why specifically to JSON. Honestly, it's the only to-string serialization we've at the moment in the app and I copied it. I doubt there's a difference with others for this small data type. Confy required values to implement `serde::ser::Serialize`, though I've not looked what kind of serialization it uses.

Comment (ivnsch) [reply src/preferences.rs:169]: Maybe we can use a binary serializer and store as blob. Looks like something we can come back to later. We definitely have to keep an eye on the preferences performance, as we may not always access them from background threads (though I don't see a use case atm to call FFI from the main thread).

**Pull request merged**